articles/sorting-algorithms.md

```diff
@@ -0,0 +1,145 @@
+
+We start with a gap of `n / 2`, where n is the length of the list. Then, we compare the first element in the list with the element `gap` elements along the list. If it's bigger, we swap them. Then we iterate through the list comparing every element to other elements `gap` distance away.
+
+Once we've iterated through the list, we choose another gap: in this example, to keep it simple, I'm just halving and then flooring the gap (so if it was 15, it would become 7). We iterate through the list again, swapping elements when necessary, and if they're swapped we compare them with the element `gap` distance _below_ them so see if a swap is necessary: effectively, we're running insertion sort on a subset of the list where the subset is every `gap` elements of the list.
+
+This can be faster than just using insertion sort, as it means that elements that are far away from where they need to be don't have to be swapped down one element at a time. It does come with added complexity, though.
+
+<svg class="sorting-demo" id="shell" height="300" width="800" viewBox="0 0 800 300"></svg>
+
+----
+
+## QuickSort
+
+QuickSort is probably the most efficient algorithm in this article for most cases. Like merge sort, it is a divide and conquer algorithm, but it works differently: it chooses a "pivot" element, and then orders the list so that all the elements smaller than the pivot are below the pivot, and all the elements larger than the pivot are above the pivot (the pivot is then in the correct place). It then runs quicksort on the list below and the list above, choosing new pivots in those lists and running itself again.
+
+The algorithm works by choosing a pivot element in the list it wants to sort—in this case, by taking a random element—and then setting `l` and `r` variables to the top and bottom points of the list. Then, `l` and `r` move towards the pivot, swapping around if `l` is larger than `r` to ensure that elements on the left of the pivot are always smaller than the pivot. When `l` or `r` gets the pivot, the pivot starts moving too, and when they are all in the same place, it means that the pivot is now in the correct position: everything to the left of it is smaller than it, and everything to the right of it is larger than it.
+
+There are more efficient ways to choose a pivot elements than just choosing a random element—for example, given that I know this is a list of unique elements, I could just choose the midpoint—but that's beyond the scope of this article.
+
+The dark blue elements are elements that have been marked as sorted, the green element is the pivot element, and the orange elements are the high and low elements.
+
+<svg class="sorting-demo" id="quick" height="300" width="800" viewBox="0 0 800 300"></svg>
+
+----
+
+## Heap sort
+
+Heap sort is probably the most complicated of all the algorithms explained in this article, as it relies on some other concepts from Computer Science—heaps.
```

Comment:
it's an em dash, github just doesn't display it like that! it's wide and designed to be used like that—like this.